File: go-backend/internal/service/report_service.go
// Package service 报告服务层
// @author <a href="https://codefather.cn">编程导航学习圈</a>
package service

import (
	"ai-test-go/internal/model"
	"ai-test-go/internal/model/vo"
	"ai-test-go/internal/repository"
	"ai-test-go/pkg/common"
	"ai-test-go/pkg/logger"
	"encoding/json"
	"errors"
	"math"

	"gorm.io/gorm"
)

// 雷达图维度名称
var radarDimensions = []string{"准确性", "完整性", "速度", "成本效率", "用户满意度"}

// 速度标准化基准值（毫秒）
const speedNormalizeBase = 10000.0

// 成本标准化系数
const costNormalizeFactor = 100.0

// 完整性计算中输出长度的除数
const completenessLengthDivisor = 1000.0

// 完整性计算中输出Token的除数
const completenessTokenDivisor = 100.0

// 标准化分数的最大值
const normalizeMaxScore = 100.0

// 成本标准化偏移量（避免除零）
const costNormalizeOffset = 0.01

// 平均成本保留小数位数
const avgCostScale = 6

type ReportService struct {
	testTaskRepo   *repository.TestTaskRepository
	testResultRepo *repository.TestResultRepository
}

func NewReportService(
	testTaskRepo *repository.TestTaskRepository,
	testResultRepo *repository.TestResultRepository,
) *ReportService {
	return &ReportService{
		testTaskRepo:   testTaskRepo,
		testResultRepo: testResultRepo,
	}
}

// GenerateReport 生成测试报告
func (s *ReportService) GenerateReport(taskID string, userID int64) (*vo.ReportVO, error) {
	task, err := s.testTaskRepo.FindByID(taskID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, common.NewBusinessException(common.NOT_FOUND_ERROR, "任务不存在")
		}
		logger.Log.Errorf("查询任务失败: taskId=%s, error=%v", taskID, err)
		return nil, common.NewBusinessException(common.SYSTEM_ERROR, "查询任务失败")
	}

	if task.UserID != userID {
		return nil, common.NewBusinessException(common.NO_AUTH_ERROR, "无权限查看该任务报告")
	}

	testResults, err := s.testResultRepo.ListByTaskID(taskID)
	if err != nil {
		logger.Log.Errorf("查询测试结果失败: taskId=%s, error=%v", taskID, err)
		return nil, common.NewBusinessException(common.SYSTEM_ERROR, "查询测试结果失败")
	}

	if len(testResults) == 0 {
		return nil, common.NewBusinessException(common.NOT_FOUND_ERROR, "该任务暂无测试结果")
	}

	report := &vo.ReportVO{
		TaskID:   taskID,
		TaskName: task.Name,
	}

	summary := s.calculateSummary(testResults)
	report.Summary = summary

	modelStatistics := s.calculateModelStatistics(testResults)
	report.ModelStatistics = modelStatistics

	radarChart := s.generateRadarChartData(testResults, modelStatistics)
	report.RadarChart = radarChart

	barChart := s.generateBarChartData(modelStatistics)
	report.BarChart = barChart

	testResultVOs := s.convertToTestResultVOs(testResults)
	report.TestResults = testResultVOs

	return report, nil
}

// calculateSummary 计算报告摘要
func (s *ReportService) calculateSummary(testResults []model.TestResult) *vo.ReportSummaryVO {
	summary := &vo.ReportSummaryVO{}

	var totalCost float64
	var totalResponseTime int64
	var responseTimeCount int
	var totalTokens int64
	modelNames := make(map[string]struct{})

	for _, result := range testResults {
		totalCost += result.Cost

		if result.ResponseTimeMs > 0 {
			totalResponseTime += int64(result.ResponseTimeMs)
			responseTimeCount++
		}

		totalTokens += int64(result.InputTokens) + int64(result.OutputTokens)

		modelNames[result.ModelName] = struct{}{}
	}

	summary.TotalCost = totalCost
	summary.TotalTokens = totalTokens
	summary.TotalResults = len(testResults)
	summary.ModelCount = len(modelNames)

	if responseTimeCount > 0 {
		avg := float64(totalResponseTime) / float64(responseTimeCount)
		summary.AvgResponseTimeMs = &avg
	}

	return summary
}

// calculateModelStatistics 计算各模型统计信息
func (s *ReportService) calculateModelStatistics(testResults []model.TestResult) []vo.ModelStatisticsVO {
	groupedByModel := make(map[string][]model.TestResult)
	for _, result := range testResults {
		groupedByModel[result.ModelName] = append(groupedByModel[result.ModelName], result)
	}

	var statisticsList []vo.ModelStatisticsVO

	for modelName, modelResults := range groupedByModel {
		statistics := vo.ModelStatisticsVO{
			ModelName: modelName,
			TestCount: len(modelResults),
		}

		var totalResponseTime int64
		var responseTimeCount int
		var totalInputTokens int64
		var inputTokensCount int
		var totalOutputTokens int64
		var outputTokensCount int
		var totalTokens int64
		var totalCost float64
		var totalUserRating int64
		var userRatingCount int
		var totalAiScore float64
		var aiScoreCount int

		for _, result := range modelResults {
			if result.ResponseTimeMs > 0 {
				totalResponseTime += int64(result.ResponseTimeMs)
				responseTimeCount++
			}

			if result.InputTokens > 0 {
				totalInputTokens += int64(result.InputTokens)
				inputTokensCount++
			}

			if result.OutputTokens > 0 {
				totalOutputTokens += int64(result.OutputTokens)
				outputTokensCount++
			}

			totalTokens += int64(result.InputTokens) + int64(result.OutputTokens)
			totalCost += result.Cost

			if result.UserRating != nil && *result.UserRating > 0 {
				totalUserRating += int64(*result.UserRating)
				userRatingCount++
			}

			if result.AIScore != "" {
				score := s.parseAiScoreTotal(result.AIScore)
				if score > 0 {
					totalAiScore += score
					aiScoreCount++
				}
			}
		}

		if responseTimeCount > 0 {
			avg := float64(totalResponseTime) / float64(responseTimeCount)
			statistics.AvgResponseTimeMs = &avg
		}

		if inputTokensCount > 0 {
			avg := float64(totalInputTokens) / float64(inputTokensCount)
			statistics.AvgInputTokens = &avg
		}

		if outputTokensCount > 0 {
			avg := float64(totalOutputTokens) / float64(outputTokensCount)
			statistics.AvgOutputTokens = &avg
		}

		statistics.TotalTokens = totalTokens
		statistics.TotalCost = totalCost

		if statistics.TestCount > 0 {
			avgCost := totalCost / float64(statistics.TestCount)
			rounded := math.Round(avgCost*math.Pow10(avgCostScale)) / math.Pow10(avgCostScale)
			statistics.AvgCost = &rounded
		}

		if userRatingCount > 0 {
			avg := float64(totalUserRating) / float64(userRatingCount)
			statistics.AvgUserRating = &avg
		}

		if aiScoreCount > 0 {
			avg := totalAiScore / float64(aiScoreCount)
			statistics.AvgAiScore = &avg
		}

		statisticsList = append(statisticsList, statistics)
	}

	return statisticsList
}

// generateRadarChartData 生成雷达图数据
func (s *ReportService) generateRadarChartData(testResults []model.TestResult, modelStatistics []vo.ModelStatisticsVO) *vo.RadarChartDataVO {
	radarChart := &vo.RadarChartDataVO{
		Dimensions: radarDimensions,
	}

	var seriesList []vo.RadarSeriesVO

	for _, statistics := range modelStatistics {
		series := vo.RadarSeriesVO{
			ModelName: statistics.ModelName,
		}

		var values []float64

		// 准确性（基于AI评分）
		accuracy := 0.0
		if statistics.AvgAiScore != nil {
			accuracy = *statistics.AvgAiScore
		}
		values = append(values, normalizeScore(accuracy, 0.0, normalizeMaxScore))

		// 完整性（基于输出文本长度和Token数）
		completeness := s.calculateCompleteness(testResults, statistics.ModelName)
		values = append(values, normalizeScore(completeness, 0.0, normalizeMaxScore))

		// 速度（响应时间越短分数越高）
		speed := 0.0
		if statistics.AvgResponseTimeMs != nil {
			speed = normalizeSpeed(*statistics.AvgResponseTimeMs)
		}
		values = append(values, speed)

		// 成本效率（成本越低分数越高）
		costEfficiency := 0.0
		if statistics.AvgCost != nil {
			costEfficiency = normalizeCostEfficiency(*statistics.AvgCost)
		}
		values = append(values, costEfficiency)

		// 用户满意度（1-5分标准化到0-100）
		userSatisfaction := 0.0
		if statistics.AvgUserRating != nil {
			userSatisfaction = normalizeScore(*statistics.AvgUserRating, 1.0, 5.0)
		}
		values = append(values, userSatisfaction)

		series.Values = values
		seriesList = append(seriesList, series)
	}

	radarChart.Series = seriesList
	return radarChart
}

// calculateCompleteness 计算完整性（基于输出文本长度和Token数）
func (s *ReportService) calculateCompleteness(testResults []model.TestResult, modelName string) float64 {
	var modelResults []model.TestResult
	for _, result := range testResults {
		if result.ModelName == modelName {
			modelResults = append(modelResults, result)
		}
	}

	if len(modelResults) == 0 {
		return 0.0
	}

	var totalOutputLength int
	var outputLengthCount int
	var totalOutputTokens int64
	var outputTokensCount int

	for _, result := range modelResults {
		if result.OutputText != "" {
			totalOutputLength += len(result.OutputText)
			outputLengthCount++
		}
		if result.OutputTokens > 0 {
			totalOutputTokens += int64(result.OutputTokens)
			outputTokensCount++
		}
	}

	avgOutputLength := 0.0
	if outputLengthCount > 0 {
		avgOutputLength = float64(totalOutputLength) / float64(outputLengthCount)
	}

	avgOutputTokens := 0.0
	if outputTokensCount > 0 {
		avgOutputTokens = float64(totalOutputTokens) / float64(outputTokensCount)
	}

	completeness := (avgOutputLength/completenessLengthDivisor + avgOutputTokens/completenessTokenDivisor) / 2.0
	return math.Min(completeness, normalizeMaxScore)
}

// normalizeScore 标准化分数到0-100范围
func normalizeScore(value, min, max float64) float64 {
	if max == min {
		return 0.0
	}
	normalized := ((value - min) / (max - min)) * normalizeMaxScore
	return math.Max(0.0, math.Min(normalizeMaxScore, normalized))
}

// normalizeSpeed 标准化速度（响应时间越短，分数越高）
func normalizeSpeed(responseTimeMs float64) float64 {
	if responseTimeMs <= 0 {
		return 0.0
	}
	normalized := speedNormalizeBase / responseTimeMs
	return math.Min(normalizeMaxScore, normalized)
}

// normalizeCostEfficiency 标准化成本效率（成本越低，分数越高）
func normalizeCostEfficiency(avgCost float64) float64 {
	if avgCost <= 0 {
		return 0.0
	}
	normalized := 1.0 / (avgCost*costNormalizeFactor + costNormalizeOffset)
	return math.Min(normalizeMaxScore, normalized*normalizeMaxScore)
}

// generateBarChartData 生成柱状图数据
func (s *ReportService) generateBarChartData(modelStatistics []vo.ModelStatisticsVO) *vo.BarChartDataVO {
	barChart := &vo.BarChartDataVO{}

	var categories []string
	for _, statistics := range modelStatistics {
		categories = append(categories, statistics.ModelName)
	}
	barChart.Categories = categories

	var seriesList []vo.BarSeriesVO

	// 平均响应时间
	responseTimeSeries := vo.BarSeriesVO{
		Name: "平均响应时间",
		Unit: "ms",
	}
	var responseTimeData []float64
	for _, statistics := range modelStatistics {
		if statistics.AvgResponseTimeMs != nil {
			responseTimeData = append(responseTimeData, *statistics.AvgResponseTimeMs)
		} else {
			responseTimeData = append(responseTimeData, 0.0)
		}
	}
	responseTimeSeries.Data = responseTimeData
	seriesList = append(seriesList, responseTimeSeries)

	// 总Token消耗
	totalTokensSeries := vo.BarSeriesVO{
		Name: "总Token消耗",
		Unit: "tokens",
	}
	var totalTokensData []float64
	for _, statistics := range modelStatistics {
		totalTokensData = append(totalTokensData, float64(statistics.TotalTokens))
	}
	totalTokensSeries.Data = totalTokensData
	seriesList = append(seriesList, totalTokensSeries)

	// 总成本
	totalCostSeries := vo.BarSeriesVO{
		Name: "总成本",
		Unit: "USD",
	}
	var totalCostData []float64
	for _, statistics := range modelStatistics {
		totalCostData = append(totalCostData, statistics.TotalCost)
	}
	totalCostSeries.Data = totalCostData
	seriesList = append(seriesList, totalCostSeries)

	barChart.Series = seriesList
	return barChart
}

// convertToTestResultVOs 批量转换测试结果为VO
func (s *ReportService) convertToTestResultVOs(testResults []model.TestResult) []vo.TestResultVO {
	var vos []vo.TestResultVO
	for _, result := range testResults {
		resultVO := vo.TestResultVO{
			ID:             result.ID,
			TaskID:         result.TaskID,
			SceneID:        result.SceneID,
			PromptID:       result.PromptID,
			ModelName:      result.ModelName,
			InputPrompt:    result.InputPrompt,
			OutputText:     result.OutputText,
			Reasoning:      result.Reasoning,
			ResponseTimeMs: result.ResponseTimeMs,
			InputTokens:    result.InputTokens,
			OutputTokens:   result.OutputTokens,
			Cost:           result.Cost,
			UserRating:     result.UserRating,
			AIScore:        result.AIScore,
			CreateTime:     result.CreateTime,
		}
		vos = append(vos, resultVO)
	}
	return vos
}

// parseAiScoreTotal 解析AI评分JSON，优先使用 averageRating（多评委），兼容 total 字段
func (s *ReportService) parseAiScoreTotal(aiScoreJSON string) float64 {
	if aiScoreJSON == "" {
		return 0.0
	}

	var aiScoreMap map[string]interface{}
	if err := json.Unmarshal([]byte(aiScoreJSON), &aiScoreMap); err != nil {
		logger.Log.Warnf("解析AI评分失败: %s, error=%v", aiScoreJSON, err)
		return 0.0
	}

	if v := parseNumberFromMap(aiScoreMap, "averageRating"); v > 0 {
		return v
	}
	return parseNumberFromMap(aiScoreMap, "total")
}

func parseNumberFromMap(m map[string]interface{}, key string) float64 {
	obj, ok := m[key]
	if !ok {
		return 0.0
	}
	switch v := obj.(type) {
	case float64:
		return v
	case int:
		return float64(v)
	case int64:
		return float64(v)
	default:
		return 0.0
	}
}

File: python-backend/app/services/report_service.py
"""
报告服务层：生成多维度对比报告、数据统计、雷达图/柱状图数据
@author <a href="https://codefather.cn">编程导航学习圈</a>
"""
import json
import logging
from collections import defaultdict
from typing import List, Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.errors import BusinessException, ErrorCode
from app.models.test_result import TestResult
from app.models.test_task import TestTask
from app.schemas.report import (
    ReportVO,
    ReportSummaryVO,
    ModelStatisticsVO,
    RadarChartDataVO,
    RadarSeriesVO,
    BarChartDataVO,
    BarSeriesVO,
    TestResultVO,
)

logger = logging.getLogger(__name__)

RADAR_DIMENSIONS = ["准确性", "完整性", "速度", "成本效率", "用户满意度"]
SPEED_NORMALIZE_DIVISOR = 10000.0
COST_EFFICIENCY_FACTOR = 100.0
COST_EFFICIENCY_OFFSET = 0.01
COMPLETENESS_MAX = 100.0
SCORE_MAX = 100.0


class ReportService:
    """
    报告服务：生成测试报告（摘要、模型统计、雷达图、柱状图、详细结果）
    """

    @staticmethod
    async def generate_report(db: AsyncSession, task_id: str, user_id: int) -> ReportVO:
        """
        生成测试报告

        Args:
            db: 数据库会话
            task_id: 任务ID
            user_id: 当前用户ID

        Returns:
            ReportVO 报告数据

        Raises:
            BusinessException: 任务不存在、无权限、无测试结果
        """
        if not task_id or not str(task_id).strip():
            raise BusinessException(ErrorCode.PARAMS_ERROR, "任务ID不能为空")

        task_result = await db.execute(
            select(TestTask).where(TestTask.id == task_id.strip(), TestTask.is_delete == 0)
        )
        task = task_result.scalar_one_or_none()
        if not task:
            raise BusinessException(ErrorCode.NOT_FOUND_ERROR, "任务不存在")
        if task.user_id != user_id:
            raise BusinessException(ErrorCode.NO_AUTH_ERROR, "无权限查看该任务报告")

        results_result = await db.execute(
            select(TestResult)
            .where(TestResult.task_id == task_id.strip(), TestResult.is_delete == 0)
            .order_by(TestResult.create_time.asc())
        )
        test_results: List[TestResult] = list(results_result.scalars().all())

        if not test_results:
            raise BusinessException(ErrorCode.NOT_FOUND_ERROR, "该任务暂无测试结果")

        summary = ReportService._calculate_summary(test_results)
        model_statistics = ReportService._calculate_model_statistics(test_results)
        radar_chart = ReportService._generate_radar_chart(test_results, model_statistics)
        bar_chart = ReportService._generate_bar_chart(model_statistics)
        test_result_vo_list = [ReportService._to_test_result_vo(r) for r in test_results]

        return ReportVO(
            task_id=task.id,
            task_name=task.name,
            summary=summary,
            model_statistics=model_statistics,
            radar_chart=radar_chart,
            bar_chart=bar_chart,
            test_results=test_result_vo_list,
        )

    @staticmethod
    def _calculate_summary(test_results: List[TestResult]) -> ReportSummaryVO:
        """计算报告摘要"""
        total_cost_float = sum(
            float(r.cost) for r in test_results if r.cost is not None
        ) or None

        response_times = [r.response_time_ms for r in test_results if r.response_time_ms is not None]
        avg_response_ms = sum(response_times) / len(response_times) if response_times else None

        total_tokens = sum(
            (r.input_tokens or 0) + (r.output_tokens or 0)
            for r in test_results
            if (r.input_tokens is not None or r.output_tokens is not None)
        )

        model_names = {r.model_name for r in test_results}
        return ReportSummaryVO(
            total_cost=total_cost_float,
            avg_response_time_ms=avg_response_ms,
            total_tokens=total_tokens,
            total_results=len(test_results),
            model_count=len(model_names),
        )

    @staticmethod
    def _calculate_model_statistics(test_results: List[TestResult]) -> List[ModelStatisticsVO]:
        """按模型分组计算统计信息"""
        grouped: dict[str, List[TestResult]] = defaultdict(list)
        for r in test_results:
            grouped[r.model_name].append(r)

        statistics_list: List[ModelStatisticsVO] = []
        for model_name, model_results in grouped.items():
            count = len(model_results)
            response_times = [r.response_time_ms for r in model_results if r.response_time_ms is not None]
            avg_response_ms = sum(response_times) / len(response_times) if response_times else None

            input_tokens = [r.input_tokens for r in model_results if r.input_tokens is not None]
            avg_input = sum(input_tokens) / len(input_tokens) if input_tokens else None
            output_tokens = [r.output_tokens for r in model_results if r.output_tokens is not None]
            avg_output = sum(output_tokens) / len(output_tokens) if output_tokens else None

            total_tokens = sum(
                (r.input_tokens or 0) + (r.output_tokens or 0)
                for r in model_results
            )
            total_cost_float = sum(
                float(r.cost) for r in model_results if r.cost is not None
            )
            avg_cost = total_cost_float / count if count else None

            user_ratings = [r.user_rating for r in model_results if r.user_rating is not None]
            avg_user_rating = sum(user_ratings) / len(user_ratings) if user_ratings else None

            ai_scores: List[float] = []
            for r in model_results:
                if not r.ai_score or not str(r.ai_score).strip():
                    continue
                try:
                    obj = json.loads(r.ai_score) if isinstance(r.ai_score, str) else r.ai_score
                    total_val = obj.get("total")
                    if total_val is not None and isinstance(total_val, (int, float)):
                        if float(total_val) > 0:
                            ai_scores.append(float(total_val))
                except (json.JSONDecodeError, TypeError) as e:
                    logger.warning("解析AI评分失败: %s, %s", r.ai_score, e)
            avg_ai_score = sum(ai_scores) / len(ai_scores) if ai_scores else None

            statistics_list.append(
                ModelStatisticsVO(
                    model_name=model_name,
                    test_count=count,
                    avg_response_time_ms=avg_response_ms,
                    avg_input_tokens=avg_input,
                    avg_output_tokens=avg_output,
                    total_tokens=total_tokens,
                    total_cost=total_cost_float if total_cost_float else None,
                    avg_cost=avg_cost,
                    avg_user_rating=avg_user_rating,
                    avg_ai_score=avg_ai_score,
                )
            )
        return statistics_list

    @staticmethod
    def _generate_radar_chart(
        test_results: List[TestResult],
        model_statistics: List[ModelStatisticsVO],
    ) -> RadarChartDataVO:
        """生成雷达图数据（准确性、完整性、速度、成本效率、用户满意度）"""
        series_list: List[RadarSeriesVO] = []
        for stat in model_statistics:
            accuracy = stat.avg_ai_score if stat.avg_ai_score is not None else 0.0
            accuracy_norm = ReportService._normalize_score(accuracy, 0.0, 100.0)

            completeness = ReportService._calculate_completeness(test_results, stat.model_name)
            completeness_norm = ReportService._normalize_score(completeness, 0.0, COMPLETENESS_MAX)

            speed = 0.0
            if stat.avg_response_time_ms is not None and stat.avg_response_time_ms > 0:
                speed = min(SCORE_MAX, SPEED_NORMALIZE_DIVISOR / stat.avg_response_time_ms)

            cost_eff = 0.0
            if stat.avg_cost is not None and stat.avg_cost > 0:
                cost_eff = min(
                    SCORE_MAX,
                    (1.0 / (stat.avg_cost * COST_EFFICIENCY_FACTOR + COST_EFFICIENCY_OFFSET)) * 100.0,
                )

            user_sat = 0.0
            if stat.avg_user_rating is not None:
                user_sat = ReportService._normalize_score(stat.avg_user_rating, 1.0, 5.0)

            series_list.append(
                RadarSeriesVO(
                    model_name=stat.model_name,
                    values=[accuracy_norm, completeness_norm, speed, cost_eff, user_sat],
                )
            )
        return RadarChartDataVO(dimensions=RADAR_DIMENSIONS, series=series_list)

    @staticmethod
    def _calculate_completeness(test_results: List[TestResult], model_name: str) -> float:
        """基于输出文本长度和 Token 数计算完整性得分"""
        model_results = [r for r in test_results if r.model_name == model_name]
        if not model_results:
            return 0.0
        lengths = [len(r.output_text or "") for r in model_results if r.output_text is not None]
        avg_len = sum(lengths) / len(lengths) if lengths else 0.0
        tokens = [r.output_tokens for r in model_results if r.output_tokens is not None]
        avg_tokens = sum(tokens) / len(tokens) if tokens else 0.0
        completeness = (avg_len / 1000.0 + avg_tokens / 100.0) / 2.0
        return min(completeness, COMPLETENESS_MAX)

    @staticmethod
    def _normalize_score(value: float, min_val: float, max_val: float) -> float:
        """将分数标准化到 0-100"""
        if max_val == min_val:
            return 0.0
        normalized = ((value - min_val) / (max_val - min_val)) * 100.0
        return max(0.0, min(SCORE_MAX, normalized))

    @staticmethod
    def _generate_bar_chart(model_statistics: List[ModelStatisticsVO]) -> BarChartDataVO:
        """生成柱状图数据（平均响应时间、总Token、总成本）"""
        categories = [s.model_name for s in model_statistics]
        response_times = [
            s.avg_response_time_ms if s.avg_response_time_ms is not None else 0.0
            for s in model_statistics
        ]
        total_tokens = [
            float(s.total_tokens) if s.total_tokens is not None else 0.0
            for s in model_statistics
        ]
        total_costs = [
            s.total_cost if s.total_cost is not None else 0.0
            for s in model_statistics
        ]
        series_list = [
            BarSeriesVO(name="平均响应时间", data=response_times, unit="ms"),
            BarSeriesVO(name="总Token消耗", data=total_tokens, unit="tokens"),
            BarSeriesVO(name="总成本", data=total_costs, unit="USD"),
        ]
        return BarChartDataVO(categories=categories, series=series_list)

    @staticmethod
    def _to_test_result_vo(r: TestResult) -> TestResultVO:
        """TestResult 转 TestResultVO"""
        cost_float = float(r.cost) if r.cost is not None else None
        ai_score_str = r.ai_score if isinstance(r.ai_score, str) else (json.dumps(r.ai_score) if r.ai_score else None)
        create_str = r.create_time.isoformat() if r.create_time else None
        return TestResultVO(
            id=r.id,
            task_id=r.task_id,
            scene_id=r.scene_id,
            prompt_id=r.prompt_id,
            model_name=r.model_name,
            input_prompt=r.input_prompt or "",
            output_text=r.output_text or "",
            reasoning=r.reasoning,
            response_time_ms=r.response_time_ms,
            input_tokens=r.input_tokens,
            output_tokens=r.output_tokens,
            cost=cost_float,
            user_rating=r.user_rating,
            ai_score=ai_score_str,
            create_time=create_str,
        )
